### PR TITLE
Write e2e test for ResourceQuota status +3 Endpoints

### LINKED ISSUE #111956: Write e2e test for ResourceQuota status +3 Endpoints
# Progress <code>[6/6]</code>

-   [X] APISnoop org-flow: [CoreV1ResourceQuotaStatusTest.org](https://github.com/apisnoop/ticket-writing/blob/master/CoreV1ResourceQuotaStatusTest.org)
-   [X] Test approval issue: #111956
-   [X] Test PR: #111957
-   [x] Two weeks soak start date: [testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.apply.changes.to.a.resourcequota.status) 19 Oct 2022
-   [x] Two weeks soak end date: 2 Nov 2022
-   [x] Test promotion PR: #113455 

# Identifying an untested feature Using APISnoop

According to this APIsnoop query, there are still three remaining ResourceQuota status endpoints which are untested.

```sql-mode
    SELECT
      endpoint,
      path,
      kind
      FROM testing.untested_stable_endpoint
      where eligible is true
      and endpoint ilike '%ResourceQuota%'
      order by kind, endpoint desc
      limit 10;
```

```example
                    endpoint                  |                            path                             |     kind
  --------------------------------------------+-------------------------------------------------------------+---------------
   replaceCoreV1NamespacedResourceQuotaStatus | /api/v1/namespaces/{namespace}/resourcequotas/{name}/status | ResourceQuota
   readCoreV1NamespacedResourceQuotaStatus    | /api/v1/namespaces/{namespace}/resourcequotas/{name}/status | ResourceQuota
   patchCoreV1NamespacedResourceQuotaStatus   | /api/v1/namespaces/{namespace}/resourcequotas/{name}/status | ResourceQuota
  (3 rows)

```

# API Reference and feature documentation

-   [Kubernetes API Reference Docs](https://kubernetes.io/docs/reference/kubernetes-api/)
-   [Kubernetes API / Policy Resource / ResourceQuota](https://kubernetes.io/docs/reference/kubernetes-api/policy-resources/resource-quota-v1/)
-   [client-go - ResourceQuota](https://github.com/kubernetes/client-go/blob/master/kubernetes/typed/core/v1/resourcequota.go)

# Test outline

```
Feature: Test patch, replace and read ResoureQuota Status api endpoints
```

-   patchCoreV1NamespacedResourceQuotaStatus

```
Scenario: the test patches a ResourceQuota Status
  Given the e2e test has created a resource quota with a set of confirmed hard limits (cpu & memory)
  And a patched status with an updated cpu hard limit
  When the test patches the Resource Quota Status
  Then the requested action is accepted without any error
  And the newly applied status hard limit for cpu is found
```

-   replaceCoreV1NamespacedResourceQuotaStatus

```
Scenario: confirm that the replace action will apply changes to a ResourceQuota Status
  Given the e2e test has a ResourceQuota after the "patch" scenario
  And a new set of hard limits for cpu and memory have been generated
  When the test updates the ResourceQuota
  Then the requested action is accepted without any error
  And the newly applied status hard limits are both found
```

-   readCoreV1NamespacedResourceQuotaStatus

```
Scenario: the test reads a ResourceQuota Status
  Given the e2e test has a ResourceQuota after the "replace" scenario
  When the test reads the ResourceQuota Status
  Then the requested action is accepted without any error
  And the status hard limits are confirmed as unchanged
```

# E2E Test

Using a number of existing e2e test practices a new [ginkgo test](https://github.com/ii/kubernetes/blob/create-resourcequota-status-test/test/e2e/apimachinery/resource_quota.go#L990-L1075) has been created for 3 ResourceQuota Status endpoints. The e2e logs for this test are listed below.

```
[It] should apply changes to a resourcequota status
  /home/ii/go/src/k8s.io/kubernetes/test/e2e/apimachinery/resource_quota.go:990
STEP: Creating a ResourceQuota 08/22/22 14:33:27.541
Aug 22 14:33:27.548: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard cpu limit of 500m
Aug 22 14:33:27.548: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard memory limit of 500Mi
STEP: patching /status 08/22/22 14:33:27.548
Aug 22 14:33:27.554: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard cpu status of 750m
STEP: updating /status 08/22/22 14:33:27.554
Aug 22 14:33:27.561: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard cpu status of 1500m
Aug 22 14:33:27.561: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard memory status of 1000Mi
STEP: get /status 08/22/22 14:33:27.561
Aug 22 14:33:27.565: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard cpu status of 1500m
Aug 22 14:33:27.565: INFO: Resource quota "e2e-quotastatus-44qxm" reports a hard memory status of 1000Mi
```

# Verifying increase in coverage with APISnoop

## Listing endpoints hit by the new e2e test

This query shows which Resource Status endpoints are hit within a short period of running this e2e test

```sql-mode
select distinct  endpoint, right(useragent,46) AS useragent
from testing.audit_event
where endpoint ilike '%ResourceQuotaStatus'
and release_date::BIGINT > round(((EXTRACT(EPOCH FROM NOW()))::numeric)*1000,0) - 60000
and useragent like 'e2e%should%'
order by endpoint
limit 10;
```

```example
                  endpoint                  |                   useragent
--------------------------------------------+------------------------------------------------
 patchCoreV1NamespacedResourceQuotaStatus   | should apply changes to a resourcequota status
 readCoreV1NamespacedResourceQuotaStatus    | should apply changes to a resourcequota status
 replaceCoreV1NamespacedResourceQuotaStatus | should apply changes to a resourcequota status
(3 rows)

```

# Final notes

If a test with these calls gets merged, **test coverage will go up by 3 point**

This test is also created with the goal of conformance promotion.

---

/sig testing
/sig architecture
/area conformance

### PR DESCRIPTION
**Pull request by riaankleinhans**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- patchCoreV1NamespacedResourceQuotaStatus
- replaceCoreV1NamespacedResourceQuotaStatus
- readCoreV1NamespacedResourceQuotaStatus

**Which issue(s) this PR fixes:**
Fixes #111956

**Testgrid Link:** 
[testgrid-link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&width=5&graph-metrics=test-duration-minutes&include-filter-by-regex=should.apply.changes.to.a.resourcequota.status)

**Special notes for your reviewer:**
Adds +3 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

### CODE REVIEW

**Comment by Riaankl** [comment]
/test pull-kubernetes-unit

Seem unrealted flake
```
=== Failed
=== FAIL: pkg/api/testing TestRoundTripTypes (5.11s)
    roundtrip.go:135: starting group ""
    roundtrip.go:254: 	round tripping to /v1, Kind=Event v1.Event
    roundtrip.go:254: 	round tripping to /v1, Kind=Event v1.Event
    roundtrip.go:254: 	round tripping to /v1, Kind=Event v1.Event
    roundtrip.go:254: 	round tripping to /v1, Kind=Event v1.Event
    ```

**Comment by dims** [comment]
note from Derek: let's make the values same as what would have been done by the controller ( https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/resourcequota/resource_quota_controller.go#L359 )

**Comment by heyste** [comment]
@derekwaynecarr  I've refocused the test to make it easier/cleaner to check the resourcequota status after "updates" or "patches" while still using the 3 outstanding api endpoints.
```
 replaceCoreV1NamespacedResourceQuotaStatus
 readCoreV1NamespacedResourceQuotaStatus
 patchCoreV1NamespacedResourceQuotaStatus
```

**Comment by derekwaynecarr** [comment]
this looks good now that the spec/status values match and should not conflict with resource quota controller itself.

/lgtm
/approve

**Comment by Riaankl** [comment]
/assign @deads2k 
for `/Approve` please.

File: test/e2e/apimachinery/resource_quota.go [hunk 1]
```diff
@@ -981,6 +988,87 @@ var _ = SIGDescribe("ResourceQuota", func() {
...
+		hardLimits = quota.Add(v1.ResourceList{}, xResourceQuota.Spec.Hard)
+
+		rqStatusJSON, err := json.Marshal(hardLimits)
+		framework.ExpectNoError(err)
+		patchedResourceQuota, err := rqClient.Patch(context.TODO(), rqName, types.StrategicMergePatchType,
+			[]byte(`{"status":`+string(rqStatusJSON)+`}`),
+			metav1.PatchOptions{}, "status")
+		framework.ExpectNoError(err)
+
+		err = wait.PollImmediate(1*time.Second, 30*time.Second, checkResourceQuotaStatus(f, *patchedResourceQuota, hardLimits))
```
Comment (deads2k) [inline test/e2e/apimachinery/resource_quota.go:1056]: A few questions

1. When you do this check, how do you know that that this e2e test made the change as opposed to the controller?  
2. Does the distinction matter to you?  
3. If you need to be sure that the /status endpoint works, do you do this for other object types by connecting a watch to the resource and looking for the value you updated?
4. Are you verifying that /status properly prevents updates to spec somewhere?

Comment (heyste) [reply test/e2e/apimachinery/resource_quota.go:1056]: I've used watches for other e2e tests for `/status` endpoints in the past but found that they could flake due to missing a watch event (you're not guaranteed to see every watch event). To work around that issue I've had a test not fail on missing a watch event. Switching to `wait.PollImmediate` was a way to have a more reliable test with better guarantees.

I'd not considered the scenario in question 4 as I my focus has been from the list of untested endpoints.

**Comment by deads2k** [comment]
This seems unlikely to flake, but also unlikely to actually test the functionality works.  I'm ok with if you really want it.

/approve
/hold

hold in case you want to handle https://github.com/kubernetes/kubernetes/pull/111957#discussion_r979044021.  A watch set up before, then handling after, with a re-attempt if the watch is disconnected seems unlikely to break.

**Comment by Riaankl** [comment]
We want to merge high quality code that meets the expectations of conformance tests.
@smarterclayton What is your thoughts?

**Comment by heyste** [comment]
@deads2k I'm looking at switching the e2e test to use `watchtools.UntilWithSync`. Also verifying that changes to `/status` prevents changes to spec.

**Comment by heyste** [comment]
@deads2k Switched the e2e test to use watches. Also, AIU it confirms that a direct update of /status can't update the spec.

**Comment by deads2k** [comment]
Significant improvement, thanks for taking the effort to do this. There may still be some possible improvements if you wanted to handle all endpoints like this, but this looks a lot better to me.

/lgtm
/approve
/hold cancel

**Pull request merged**